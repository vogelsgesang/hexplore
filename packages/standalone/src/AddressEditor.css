.hv-address-editor {
    border: 1px solid transparent;
    background: transparent;
}

.hv-address-editor:focus, .hv-address-editor:hover {
    border-color: #ccc;
    background: white;
}