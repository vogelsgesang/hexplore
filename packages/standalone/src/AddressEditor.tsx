import React, {useState, useRef, forwardRef, useImperativeHandle, useCallback} from "react";

import "./AddressEditor.css";

interface AdressEditorProperties {
    address: number;
    setAddress: (a: number) => void;
}

function parseAddress(str: string) {
    str = str.trim();
    if (/^0x[0-9a-fA-F]+$/.test(str)) {
        return Number.parseInt(str.substr(2), 16);
    } else if (/^[0-9]+$/.test(str)) {
        return Number.parseInt(str, 10);
    }
    return undefined;
}

export const AddressEditor = forwardRef(function AddressEditor({address, setAddress}: AdressEditorProperties, ref) {
    const [draft, setDraft] = useState<string | undefined>(undefined);
    const inputRef = useRef<HTMLInputElement>(null);

    const addressValue = "0x" + address.toString(16);
    const displayedValue = draft ?? addressValue;

    function onKeyDown(e: React.KeyboardEvent) {
        if (!inputRef.current) return;
        if (e.key == "Enter") {
            let parsed: number | undefined = address;
            if (draft !== undefined) {
                parsed = parseAddress(draft);
            }
            if (parsed !== undefined) {
                setAddress(parsed);
                inputRef.current?.blur();
            }
            e.preventDefault();
        } else if (e.key == "Escape") {
            setDraft(addressValue);
            // Must set the "value" synchronously, such that the "setSelectionRange" works properly
            inputRef.current.value = addressValue;
            inputRef.current.setSelectionRange(0, inputRef.current.value.length);
            e.preventDefault();
        }
    }

    const onFocus = useCallback(() => {
        inputRef.current?.setSelectionRange(0, inputRef.current?.value.length);
    }, [inputRef]);

    useImperativeHandle(
        ref,
        () => ({
            focus: () => inputRef.current?.focus(),
        }),
        [inputRef],
    );

    return (
        <input
            className="hv-address-editor"
            type="text"
            ref={inputRef}
            value={displayedValue}
            onFocus={onFocus}
            onChange={e => setDraft(e.currentTarget.value)}
            onKeyDown={onKeyDown}
            onBlur={() => setDraft(undefined)}
            width={16}
        />
    );
});
