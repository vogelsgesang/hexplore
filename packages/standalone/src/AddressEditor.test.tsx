import React from "react";
import snapshotRenderer from "react-test-renderer";
import {render, fireEvent, screen, act} from "@testing-library/react";
import {AddressEditor} from "./AddressEditor";

describe("displays the address correctly", () => {
    function testRendering(pos: number) {
        const component = snapshotRenderer.create(<AddressEditor address={pos} setAddress={() => {}} />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    }

    test("for address 0", () => {
        testRendering(0);
    });
    test("for address 33", () => {
        testRendering(33);
    });
});

describe("can be edited", () => {
    test("selects whole text on focus", () => {
        const handleChange = jest.fn();
        render(<AddressEditor address={0x7b} setAddress={handleChange} />);
        const inputField = screen.getByDisplayValue("0x7b") as HTMLInputElement;
        act(() => {
            inputField.focus();
        });
        expect(inputField.selectionStart).toBe(0);
        expect(inputField.selectionEnd).toBe(inputField.value.length);
    });
    test("sets address on enter", () => {
        const handleChange = jest.fn();
        render(<AddressEditor address={0x7b} setAddress={handleChange} />);
        fireEvent.change(screen.getByDisplayValue("0x7b"), {target: {value: "180"}});
        fireEvent.keyDown(screen.getByDisplayValue("180"), {key: "Enter"});
        expect(handleChange).toHaveBeenCalledWith(180);
    });
    test("does not set address during editing", () => {
        const handleChange = jest.fn();
        render(<AddressEditor address={0x7b} setAddress={handleChange} />);
        fireEvent.change(screen.getByDisplayValue("0x7b"), {target: {value: "180"}});
        // No "Enter" key has been hit -> no callback was called
        expect(handleChange).not.toHaveBeenCalled();
    });
    test("does not accept invalid strings", () => {
        const handleChange = jest.fn();
        render(<AddressEditor address={0x7b} setAddress={handleChange} />);
        fireEvent.change(screen.getByDisplayValue("0x7b"), {target: {value: "not an address"}});
        fireEvent.keyDown(screen.getByDisplayValue("not an address"), {key: "Enter"});
        expect(handleChange).not.toHaveBeenCalled();
    });
    test("resets address on escape and selects whole text", () => {
        const handleChange = jest.fn();
        render(<AddressEditor address={0x7b} setAddress={handleChange} />);
        const inputField = screen.getByDisplayValue("0x7b") as HTMLInputElement;
        inputField.value = "180";
        fireEvent.change(inputField, {target: {value: "180"}});
        fireEvent.keyDown(inputField, {key: "Escape"});
        expect(inputField.value).toBe("0x7b");
        expect(inputField.selectionStart).toBe(0);
        expect(inputField.selectionEnd).toBe(inputField.value.length);
        expect(handleChange).not.toHaveBeenCalled();
    });
});
