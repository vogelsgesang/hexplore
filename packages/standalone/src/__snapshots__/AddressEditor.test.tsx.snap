// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays the address correctly for address 0 1`] = `
<input
  className="hv-address-editor"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  type="text"
  value="0x0"
  width={16}
/>
`;

exports[`displays the address correctly for address 33 1`] = `
<input
  className="hv-address-editor"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  type="text"
  value="0x21"
  width={16}
/>
`;
